File: my-folder/1146-greatest-common-divisor-of-strings/solution.java
class Solution {
    public String gcdOfStrings(String str1, String str2) {
        
        if (str2.length() < str1.length()){
            String temp = str2;
            str2 = str1;
            str1 = temp;
        }
        for (int i=str1.length();i>=0;i--){
            if (canDivide(str1.substring(0,i), str1, str2)){
                return str1.substring(0,i);
            }
        }
        return "";
    }

    public boolean canDivide(String t, String str1, String str2){
        if (t == ""){
            return false;
        }
        return t.repeat(str1.length()/t.length()).equals(str1) && t.repeat(str2.length()/t.length()).equals(str2);
    }
}

File: my-folder/1146-greatest-common-divisor-of-strings/solution.py
class Solution:
    def gcdOfStrings(self, str1: str, str2: str) -> str:
        

        # given 2 strings s and t
        # s = abcabc, t = abc
        # return largest string x such that x divides both s and t

        # 

        if len(str1) > len(str2):
            str1, str2 = str2, str1

        
        def can_divide(t):
            if not t: return False
            return t * (len(str1)//len(t)) == str1 and t * (len(str2)//len(t)) == str2
        
        for i in range(len(str1)-1,-1,-1):
            if can_divide(str1[:i+1]):
                return str1[:i+1]
        
        return ''



